astropy/units/physical.py

```diff
@@ -1,133 +1,461 @@
+    else:
+        raise TypeError("Expecting a unit or a physical type")
+
+
+class PhysicalType:
+    """
+    Represents and provides information on the physical type(s) that are
+    associated with a set of units.
+
+    Instances of this class should be accessed using the ``physical_type``
+    property of `Unit` instances.  This class is not intended to be
+    instantiated directly in user code.
+
+    Parameters
+    ----------
+    unit : u.Unit
+        The unit to be represented by the physical type.
+
+    physical_types : `str` or `set` of `str`
+        A `str` representing the name of the physical type of the unit,
+        or a `set` containing strings that represent one or more names
+        of physical types.
+
+    Notes
+    -----
+    A physical type will be considered equal to an equivalent
+    `PhysicalType` instance (recommended) or a string that contains a
+    name of the physical type.  The latter method is not recommended
+    in packages, as the names of some physical types may change in the
+    future.
+
+    To maintain backwards compatibility, two physical type names may be
+    included in one string if they are separated with a slash (e.g.,
+    ``"momentum/impulse"``).  String representations of physical types
+    may include underscores instead of spaces.
+
+    Examples
+    --------
+    The preferred method to access a physical type is by accessing the
+    ``physical_type`` attribute of a unit.
+
+    >>> import astropy.units as u
+    >>> u.meter.physical_type
+    'length'
+
+    Some units correspond to multiple physical types.
+
+    >>> pressure = u.Pa.physical_type
+    >>> print(pressure)
+    {'energy density', 'pressure', 'stress'}
+    >>> 'energy density' in pressure
+    True
+
+    Physical types can be tested for equality against other physical
+    type objects or against strings that may contain the name of a
+    physical type.
+
+    >>> area = (u.m ** 2).physical_type
+    >>> area == u.barn.physical_type
+    True
+    >>> area == "area"
+    True
+
+    Multiplication, division, and exponentiation are enabled so that
+    physical types may be used for dimensional analysis.
+
+    >>> length = u.pc.physical_type
+    >>> area = (u.cm ** 2).physical_type
+    >>> length * area
+    'volume'
+    >>> area / length
+    'length'
+    >>> length ** 3
+    'volume'
+
+    Unknown physical types are labelled as ``"unknown"``.
+
+    >>> (u.s ** 13).physical_type
+    'unknown'
+
+    Dimensional analysis may be performed for unknown physical types too.
+
+    >>> length_to_19th_power = (u.m ** 19).physical_type
+    >>> length_to_20th_power = (u.m ** 20).physical_type
+    >>> length_to_19th_power, length_to_20th_power
+    ('unknown', 'unknown')
+    >>> length_to_20th_power / length_to_19th_power
+    'length'
+    """
+
+    def __init__(self, unit, physical_types):
+
+        if not isinstance(unit, core.UnitBase):
+            raise TypeError(f"Expecting a unit, not {unit}.")
```

Comment:
Cover this? (Or, given that this is for internal use, just omit the test? Why stop someone who writes a Unit ducktype?)